lib/spack/spack/util/spec_set.py

```diff
@@ -0,0 +1,189 @@
+##############################################################################
+# Copyright (c) 2013-2017, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the LICENSE file for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+import itertools
+
+import llnl.util.tty as tty
+from llnl.util.tty.colify import colify
+
+import spack
+import spack.compilers
+import spack.architecture as sarch
+import spack.schema as schema
+import spack.util.spack_yaml as syaml
+
+from spack.spec import Spec, ArchSpec
+
+
+class CombinatorialSpecSet:
+    """Set of combinatorial Specs constructed from YAML file."""
+
+    def __init__(self, yaml_like, ignore_invalid=True):
+        """Construct a combinatorial Spec set.
+
+        Args:
+            yaml_like: either raw YAML data as a dict, a file-like object
+                to read the YAML from, or a string containing YAML.  * A
+                file-like object to read YAML from. In the first case, we
+                assume already-parsed YAML data.  In the second two
+                cases, we just run yaml.load() on the data.
+            ignore_invalid (bool): whether to ignore invalid specs when
+                expanding the values of this spec set.
+        """
+        self.ignore_invalid = ignore_invalid
+
+        if isinstance(yaml_like, dict):
+            # if it's raw data, just assign it to self.data
+            self.data = yaml_like
+        else:
+            # otherwise try to load it.
+            self.data = syaml.load(yaml_like)
+
+        # validate against the test suite schema
+        schema.validate(self.data, schema.test_suite.schema)
```

Comment:
@tgamblin Is this `validate` method something that exists already in some other spack subpackage (I'm not finding it in the `schema` subpackage), or was it intended to be added as a part of this PR?  I cannot find that such a thing exists at the moment, so creating one of these `CombinatorialSpecSet` objects is problematic currently.